### PR TITLE
token-swap: Add offset curve and math

### PR DESCRIPTION
**Pull request by joncinque**
In order to properly do the single token conversion for the normal constant product curve and the new offset curve, we need to calculate a square root, since the total value of the pool goes up with the square root of the token amount.

This also contains a fix for calculating the value of deposits.  Before, we were getting the proportion of the deposit based on the amount currently stored in the swap, which is incorrect.  Now it's calculated based on how much the swap *will have* after the deposit.

This resurrects the `PreciseNumber` work used for weights in #574 and some of the work removed from https://github.com/solana-labs/solana-program-library/pull/508/commits/a8ef645b866063104869238abd387694fcd97179.  In the end, the Taylor series expansion being used for the power wasn't always the right option.  When some of the amount is close to 0, we can have huge relative swings when hanging out close to 0, which makes the base of our exponent too big to reliably calculate, so instead we're using a newtonian approximation.

~~Note that this depends on the changes in #934, so those are included in this PR as well.~~

PR has been rebased on #934 for easier reading.

@CriesofCarrots since you've looked at a bit of the math before, could you look through the changes at `math.rs` and specifically `trading_tokens_to_pool_tokens` in `offset.rs` and `calculator.rs`?

### CODE REVIEW

**Comment by joncinque** [comment]
cargo-build-test and js-token-swap both pass locally

**Review by CriesofCarrots — COMMENTED** [review]
Nits in math.rs

File: token-swap/program/src/curve/math.rs [hunk 1]
```diff
@@ -39,3 +39,407 @@ impl U256 {
...
+            return Some(precise_whole);
+        }
+        let precise_remainder = self
+            .checked_pow_approximation(&remainder_exponent, Self::MAX_APPROXIMATION_ITERATIONS)?;
+        precise_whole.checked_mul(&precise_remainder)
+    }
+
+    /// Approximate the nth root of a number using Newton's method
+    /// https://en.wikipedia.org/wiki/Newton%27s_method
+    pub fn newtonian_root_approximation(&self, root: &Self, mut guess: Self) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (joncinque) [reply token-swap/program/src/curve/math.rs:300]: And surprisingly accurate!

**Review by CriesofCarrots — APPROVED** [review]
I only examined the parts requested, but those lgtm! Thanks for updating the deposit/withdrawal handling for Offset

**Comment by joncinque** [comment]
Merging on red -- tests pass locally

**Pull request merged**